File: code/java/rest/server/src/main/java/com/taskapi/controller/TaskController.java
package com.taskapi.controller;

import com.taskapi.model.*;
import com.taskapi.service.TaskService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/tasks")
@RequiredArgsConstructor
@Tag(name = "Tasks", description = "Task management endpoints")
@CrossOrigin(origins = "*")
public class TaskController {
    
    private final TaskService taskService;
    
    @GetMapping
    @Operation(summary = "List all tasks", description = "Retrieve a paginated list of tasks with optional filters")
    @ApiResponse(responseCode = "200", description = "Tasks retrieved successfully")
    public ResponseEntity<PageResponse<Task>> listTasks(
            @Parameter(description = "Number of items per page")
            @RequestParam(defaultValue = "20") int pageSize,
            
            @Parameter(description = "Token for the next page")
            @RequestParam(required = false) String pageToken,
            
            @Parameter(description = "Filter by task status")
            @RequestParam(required = false) TaskStatus status,
            
            @Parameter(description = "Filter by assigned user")
            @RequestParam(required = false) String assignedTo,
            
            @Parameter(description = "Filter by tags (comma-separated)")
            @RequestParam(required = false) String tags,
            
            @Parameter(description = "Sort order")
            @RequestParam(required = false) String sortOrder) {
        
        List<String> tagList = null;
        if (tags != null && !tags.isEmpty()) {
            tagList = Arrays.asList(tags.split(","));
        }
        
        PageResponse<Task> response = taskService.listTasks(
                Math.min(pageSize, 100),
                pageToken,
                status,
                assignedTo,
                tagList,
                sortOrder
        );
        
        return ResponseEntity.ok(response);
    }
    
    @GetMapping("/{id}")
    @Operation(summary = "Get a task", description = "Retrieve a specific task by ID")
    @ApiResponse(responseCode = "200", description = "Task retrieved successfully")
    @ApiResponse(responseCode = "404", description = "Task not found")
    public ResponseEntity<Task> getTask(@PathVariable String id) {
        return taskService.getTask(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }
    
    @PostMapping
    @Operation(summary = "Create a task", description = "Create a new task")
    @ApiResponse(responseCode = "201", description = "Task created successfully")
    @ApiResponse(responseCode = "400", description = "Invalid request data")
    public ResponseEntity<Task> createTask(@Valid @RequestBody Task task) {
        Task created = taskService.createTask(task);
        return ResponseEntity.status(HttpStatus.CREATED).body(created);
    }
    
    @PutMapping("/{id}")
    @Operation(summary = "Update a task", description = "Update an existing task")
    @ApiResponse(responseCode = "200", description = "Task updated successfully")
    @ApiResponse(responseCode = "404", description = "Task not found")
    public ResponseEntity<Task> updateTask(
            @PathVariable String id,
            @Valid @RequestBody Task task) {
        return taskService.updateTask(id, task)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }
    
    @PatchMapping("/{id}/status")
    @Operation(summary = "Update task status", description = "Update only the status of a task")
    @ApiResponse(responseCode = "200", description = "Status updated successfully")
    @ApiResponse(responseCode = "404", description = "Task not found")
    public ResponseEntity<Task> updateTaskStatus(
            @PathVariable String id,
            @RequestBody Map<String, String> statusUpdate) {
        
        try {
            TaskStatus status = TaskStatus.valueOf(statusUpdate.get("status").toUpperCase());
            return taskService.updateTaskStatus(id, status)
                    .map(ResponseEntity::ok)
                    .orElse(ResponseEntity.notFound().build());
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().build();
        }
    }
    
    @DeleteMapping("/{id}")
    @Operation(summary = "Delete a task", description = "Delete a task by ID")
    @ApiResponse(responseCode = "204", description = "Task deleted successfully")
    @ApiResponse(responseCode = "404", description = "Task not found")
    public ResponseEntity<Void> deleteTask(@PathVariable String id) {
        if (taskService.deleteTask(id)) {
            return ResponseEntity.noContent().build();
        }
        return ResponseEntity.notFound().build();
    }
}

File: code/kotlin/rest/server/src/main/kotlin/com/taskapi/controller/TaskController.kt
package com.taskapi.controller

import com.taskapi.model.*
import com.taskapi.service.TaskService
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*

fun Application.configureRouting() {
    val taskService = TaskService()

    routing {
        route("/api/v1") {
            taskRoutes(taskService)
        }
    }
}

fun Route.taskRoutes(taskService: TaskService) {
    route("/tasks") {
        get {
            try {
                val params = QueryParams(
                    pageSize = call.request.queryParameters["pageSize"]?.toIntOrNull() ?: 20,
                    pageToken = call.request.queryParameters["pageToken"],
                    status = call.request.queryParameters["status"]?.let { 
                        try { TaskStatus.valueOf(it.uppercase()) } catch (e: Exception) { null }
                    },
                    assignedTo = call.request.queryParameters["assignedTo"],
                    tags = call.request.queryParameters["tags"],
                    sortOrder = call.request.queryParameters["sortOrder"]
                )
                
                val response = taskService.listTasks(params)
                call.respond(HttpStatusCode.OK, response)
            } catch (e: Exception) {
                call.respond(
                    HttpStatusCode.InternalServerError,
                    ErrorResponse("internal_error", "An error occurred while listing tasks")
                )
            }
        }

        get("/{id}") {
            val id = call.parameters["id"]
            if (id.isNullOrBlank()) {
                call.respond(
                    HttpStatusCode.BadRequest,
                    ErrorResponse("invalid_request", "Task ID is required")
                )
                return@get
            }

            val task = taskService.getTask(id)
            if (task != null) {
                call.respond(HttpStatusCode.OK, task)
            } else {
                call.respond(
                    HttpStatusCode.NotFound,
                    ErrorResponse("not_found", "Task with ID $id not found")
                )
            }
        }

        post {
            try {
                val request = call.receive<CreateTaskRequest>()
                
                // Validate request
                if (request.title.isBlank()) {
                    call.respond(
                        HttpStatusCode.BadRequest,
                        ErrorResponse("validation_error", "Title is required")
                    )
                    return@post
                }
                
                if (request.title.length > 200) {
                    call.respond(
                        HttpStatusCode.BadRequest,
                        ErrorResponse("validation_error", "Title must be 200 characters or less")
                    )
                    return@post
                }
                
                val task = taskService.createTask(request)
                call.respond(HttpStatusCode.Created, task)
            } catch (e: Exception) {
                call.respond(
                    HttpStatusCode.BadRequest,
                    ErrorResponse("invalid_request", "Invalid request data")
                )
            }
        }

        put("/{id}") {
            val id = call.parameters["id"]
            if (id.isNullOrBlank()) {
                call.respond(
                    HttpStatusCode.BadRequest,
                    ErrorResponse("invalid_request", "Task ID is required")
                )
                return@put
            }

            try {
                val request = call.receive<UpdateTaskRequest>()
                
                // Validate if title is provided
                if (request.title != null && request.title.isBlank()) {
                    call.respond(
                        HttpStatusCode.BadRequest,
                        ErrorResponse("validation_error", "Title cannot be empty")
                    )
                    return@put
                }
                
                if (request.title != null && request.title.length > 200) {
                    call.respond(
                        HttpStatusCode.BadRequest,
                        ErrorResponse("validation_error", "Title must be 200 characters or less")
                    )
                    return@put
                }
                
                val task = taskService.updateTask(id, request)
                if (task != null) {
                    call.respond(HttpStatusCode.OK, task)
                } else {
                    call.respond(
                        HttpStatusCode.NotFound,
                        ErrorResponse("not_found", "Task with ID $id not found")
                    )
                }
            } catch (e: Exception) {
                call.respond(
                    HttpStatusCode.BadRequest,
                    ErrorResponse("invalid_request", "Invalid request data")
                )
            }
        }

        patch("/{id}/status") {
            val id = call.parameters["id"]
            if (id.isNullOrBlank()) {
                call.respond(
                    HttpStatusCode.BadRequest,
                    ErrorResponse("invalid_request", "Task ID is required")
                )
                return@patch
            }

            try {
                val request = call.receive<UpdateStatusRequest>()
                val task = taskService.updateTaskStatus(id, request.status)
                
                if (task != null) {
                    call.respond(HttpStatusCode.OK, task)
                } else {
                    call.respond(
                        HttpStatusCode.NotFound,
                        ErrorResponse("not_found", "Task with ID $id not found")
                    )
                }
            } catch (e: Exception) {
                call.respond(
                    HttpStatusCode.BadRequest,
                    ErrorResponse("invalid_request", "Invalid request data")
                )
            }
        }

        delete("/{id}") {
            val id = call.parameters["id"]
            if (id.isNullOrBlank()) {
                call.respond(
                    HttpStatusCode.BadRequest,
                    ErrorResponse("invalid_request", "Task ID is required")
                )
                return@delete
            }

            val deleted = taskService.deleteTask(id)
            if (deleted) {
                call.respond(HttpStatusCode.NoContent)
            } else {
                call.respond(
                    HttpStatusCode.NotFound,
                    ErrorResponse("not_found", "Task with ID $id not found")
                )
            }
        }
    }
}